homeassistant/components/hassmpris/media_player.py

```diff
@@ -0,0 +1,674 @@
+
+import asyncio
+import json
+import re
+from typing import Any, cast
+
+from hassmpris.proto import mpris_pb2
+import hassmpris_client
+import voluptuous as vol
+
+from homeassistant.components.media_player import (
+    MediaPlayerDeviceClass,
+    MediaPlayerEntity,
+    MediaPlayerEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    STATE_IDLE,
+    STATE_OFF,
+    STATE_PAUSED,
+    STATE_PLAYING,
+    STATE_UNKNOWN,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_validation as cv, entity_registry as er
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+import homeassistant.util.dt as dt_util
+
+from .const import (
+    ATTR_PLAYBACK_RATE,
+    DOMAIN,
+    ENTRY_CLIENT,
+    ENTRY_MEDIA_PLAYER_ENTITY_MANAGER,
+    ENTRY_UNLOADERS,
+    LOGGER as _LOGGER,
+)
+
+PLATFORM = "media_player"
+
+DISCOVERY_SCHEMA = vol.Schema(
+    {
+        vol.Required("player_id"): cv.string,
+    }
+)
+
+SUPPORTED_MINIMAL = (
+    MediaPlayerEntityFeature.PAUSE
+    | MediaPlayerEntityFeature.PLAY
+    | MediaPlayerEntityFeature.STOP
+)
+SUPPORTED_TURN_OFF = MediaPlayerEntityFeature.TURN_OFF
+SUPPORTED_TURN_ON = MediaPlayerEntityFeature.TURN_ON
+
+# Enable to remove clones (second / third, nth instances)
+# of any player when the player exits.  Not enabled by
+# default because events in the logbook disappear when
+# the player goes away.
+# Clones are always removed upon start or reload of the
+# integration, since there is no value in keeping them
+# around.
+REMOVE_CLONES_WHILE_RUNNING = False
+
+
+def _feat2bitfield(bitfield: int, obj: object) -> str:
+    fields = []
+    for feat in dir(obj):
+        val = getattr(obj, feat)
+        if not isinstance(val, int):
+            continue
+        if bitfield & val:
+            fields.append(feat)
+    return ",".join(fields)
+
+
+def _get_player_id(
+    entity: er.RegistryEntry,
+) -> str:
+    return entity.unique_id.split("-", 1)[1]
+
+
+class HASSMPRISEntity(MediaPlayerEntity):
+    """This class represents an MPRIS media player entity."""
+
+    _attr_device_class = MediaPlayerDeviceClass.TV
+    _attr_supported_features = SUPPORTED_MINIMAL
+    _attr_playback_rate: float = 1.0
+    _attr_should_poll: bool = False
+
+    def __init__(
+        self,
+        client: hassmpris_client.AsyncMPRISClient,
+        integration_id: str,
+        player_id: str,
+        initial_state: str | None = None,
+    ) -> None:
+        """
+        Initialize the entity.
+
+        Parameters:
+          client: the client to the remote agent
+          integration_id: unique identifier of the integration
+          player_id: the name / unique identifier of the player
+        """
+        super().__init__()
+        if client is None:
+            raise ValueError("Instantiation of this class requires a client")
```

Comment:
Please don't raise exception in `__init__`.  We don't want any side effects in `__init___`

Reply:
Reasonable.  Fixed.  Hang for push.